File: console/src/api/research.ts
/** 选品调研 REST 封装与纯函数（真实数据版）。 */

import { api } from "./client";

export type ResearchStatus = "running" | "succeeded" | "partial" | "failed";
export type StepKind = "plan" | "fetch" | "cross" | "synthesize" | "agent";

export interface ResearchRun {
  id: string;
  category: string;
  boards: string[];
  status: ResearchStatus;
  mode: string;
  signal_count: number | null;
  error: string | null;
  narrative: string | null;
  created_at: string | null;
  started_at: string | null;
  finished_at: string | null;
}

export interface ResearchStep {
  id: string;
  seq: number;
  kind: StepKind;
  source_id: string | null;
  args_digest: string;
  summary: string | null;
  status: "ok" | "failed" | "skipped";
  seconds: number | null;
  error: string | null;
}

export interface Evidence {
  source_id: string;
  metric: string;
  value: number;
  url: string;
  captured_at: string;
}

export interface ResearchCandidateRow {
  id: string;
  run_id?: string;
  rank: number;
  name: string;
  platform: string;
  board: string;
  price: string;
  sales_signal: string;
  heat: number;
  keywords: string[];
  score: number;
  score_basis: Record<string, unknown>;
  evidence: Evidence[];
}

export interface ResearchDetail {
  run: ResearchRun;
  steps: ResearchStep[];
  candidates: ResearchCandidateRow[];
}

export interface SourceHealthRow {
  id: string;
  name: string;
  board: string;
  tier: string;
  health: string;
  ttl_seconds: number;
  capabilities: Record<string, unknown>;
}

export const METRIC_LABEL: Record<string, string> = {
  heat: "热度",
  trend: "趋势",
  rank: "排名",
  sales_proxy: "销量代理",
  price_band: "价格带",
};

export const STEP_KIND_LABEL: Record<StepKind, string> = {
  plan: "规划",
  fetch: "取数",
  cross: "交叉验证",
  synthesize: "成文",
  agent: "叙述",
};

export function startResearch(category: string) {
  return api.post<{ run_id: string; status: ResearchStatus; cached: boolean }>("/api/research", {
    category,
  });
}

export function fetchRunDetail(runId: string) {
  return api.get<ResearchDetail>(`/api/research/runs/${runId}`);
}

export async function fetchSourceHealth(): Promise<SourceHealthRow[]> {
  const r = await api.get<{ sources: SourceHealthRow[] }>("/api/research/sources");
  return r.sources;
}

export async function fetchRuns(limit = 5): Promise<ResearchRun[]> {
  const r = await api.get<{ runs: ResearchRun[] }>("/api/research/runs");
  return r.runs.slice(0, limit);
}

/** 价格带字符串取最低价（币种无关）：¥29-59→29、$12.99→12.99、""→0。 */
export function parsePriceMin(s: string): number {
  const m = /-?\d+(?:\.\d+)?/.exec(s ?? "");
  return m ? Number(m[0]) : 0;
}

export function nextPollStatus(run: Pick<ResearchRun, "status">): "pending" | "done" {
  return run && run.status !== "running" ? "done" : "pending";
}

/** 候选 → 收藏形状；source 带 run_id 前缀便于溯源。 */
export function candidateToFavorite(c: ResearchCandidateRow): {
  name: string;
  source: string;
  heat: number;
  price: number;
} {
  const runTag = (c.run_id ?? c.id).replace(/^rsr[_-]?/, "").slice(0, 6);
  return {
    name: c.name,
    source: `rsr ${runTag} · ${c.platform}`,
    heat: c.heat,
    price: parsePriceMin(c.price),
  };
}

export function stepIconFor(status: ResearchStep["status"]): "done" | "failed" | "skipped" {
  if (status === "ok") return "done";
  if (status === "failed") return "failed";
  return "skipped";
}

File: server/ecos/models/research.py
"""研究子域 ORM：调研运行 / 步骤 / 候选 / 源快照缓存。"""

from datetime import datetime

from sqlalchemy import DateTime, Float, JSON, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base, TimestampMixin, new_id, utcnow


class ResearchRun(Base, TimestampMixin):
    __tablename__ = "research_runs"

    id: Mapped[str] = mapped_column(primary_key=True, default=lambda: new_id("rsr"))
    user_id: Mapped[str] = mapped_column(index=True)
    category: Mapped[str]
    boards: Mapped[list] = mapped_column(JSON, default=list)
    status: Mapped[str] = mapped_column(default="running")  # running|succeeded|partial|failed
    mode: Mapped[str] = mapped_column(default="rule")  # rule|agent
    signal_count: Mapped[int] = mapped_column(default=0)
    error: Mapped[str | None] = mapped_column(default=None)
    narrative: Mapped[str | None] = mapped_column(default=None)
    tokens_in: Mapped[int] = mapped_column(default=0)
    tokens_out: Mapped[int] = mapped_column(default=0)
    started_at: Mapped[datetime | None] = mapped_column(DateTime, default=None)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, default=None)


class ResearchStep(Base, TimestampMixin):
    __tablename__ = "research_steps"

    id: Mapped[str] = mapped_column(primary_key=True, default=lambda: new_id("rst"))
    run_id: Mapped[str] = mapped_column(index=True)
    seq: Mapped[int]
    kind: Mapped[str]  # plan|fetch|cross|synthesize|agent
    source_id: Mapped[str] = mapped_column(default="")
    args_digest: Mapped[str] = mapped_column(default="")
    summary: Mapped[str] = mapped_column(default="")
    status: Mapped[str] = mapped_column(default="ok")  # ok|failed|skipped
    seconds: Mapped[float] = mapped_column(Float, default=0.0)
    error: Mapped[str | None] = mapped_column(default=None)


class ResearchCandidate(Base, TimestampMixin):
    __tablename__ = "research_candidates"

    id: Mapped[str] = mapped_column(primary_key=True, default=lambda: new_id("rcd"))
    run_id: Mapped[str] = mapped_column(index=True)
    rank: Mapped[int]
    name: Mapped[str]
    platform: Mapped[str] = mapped_column(default="")
    board: Mapped[str] = mapped_column(default="")
    price: Mapped[str] = mapped_column(default="")
    sales_signal: Mapped[str] = mapped_column(default="")
    heat: Mapped[int] = mapped_column(default=0)
    keywords: Mapped[list] = mapped_column(JSON, default=list)
    score: Mapped[float] = mapped_column(Float, default=0.0)
    score_basis: Mapped[dict] = mapped_column(JSON, default=dict)
    evidence: Mapped[list] = mapped_column(JSON, default=list)


class SourceSnapshot(Base, TimestampMixin):
    __tablename__ = "research_snapshots"
    __table_args__ = (UniqueConstraint("source_id", "query_key", name="uq_snapshot_key"),)

    id: Mapped[str] = mapped_column(primary_key=True, default=lambda: new_id("rsc"))
    source_id: Mapped[str] = mapped_column(index=True)
    query_key: Mapped[str]
    payload: Mapped[list] = mapped_column(JSON, default=list)
    captured_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
    expires_at: Mapped[datetime] = mapped_column(DateTime)
